posts/2018-07-12-servant-dsl-typelevel.md

```diff
@@ -0,0 +1,664 @@
+approach to a very simplified web application description "language" that we
+could make out of tiny building blocks (static path fragments, captures, etc).
+
+# A first modular attempt
+
+Adapting the approach from the previous section to our domain, we can give a
+shot at decomposing the kind of information we want to represent into
+a few different "constructs" (i.e data types).
+
+``` haskell
+-- static path fragments
+data Static = Static String
+
+-- variable path fragments ("captures")
+data Capture = Capture
+
+-- HTTP method
+data Method = Get | Post
+-- Leaf of a chain of :>'s, specifies the HTTP method
+data Verb = Verb Method
+
+-- chain a few "endpoint components" with this operator,
+-- all chains must be terminated with a 'Verb' component.
+infixr :> 5
+data a :> b = a :> b
+
+-- a class to specify all the valid endpoint descriptions
+class Endpoint a
+instance Endpoint (Verb a)
+instance Endpoint rest => Endpoint (Static :> rest)
+instance Endpoint rest => Endpoint (Capture :> rest)
+
+-- GET /hello
+endpoint1 = Static "hello" :> Verb Get
+```
+
+OK, why not. Let's now try to write an interpretation for generating links
+to endpoints like the one above. This is a lot simpler and self-contained than
+investigating client generation or server-side routing, while retaining many of
+the difficulties. The main one is that depending on what we find in the
+description of the endpoint, we need the type of the link-generating function
+to change: indeed, if we encounter `Capture`s, then the user has to supply
+values for them. We will let the user do that through one additional
+argument per Capture we encounter.
+
+Let's start with something really simple.
+
+``` haskell
+type Link = [String]
+
+-- @renderLink ["hello", "world"] == "/hello/world"@
+renderLink :: Link -> String
+renderLink xs = '/' : intercalate '/' xs
+
+class HasLink endpoint where
+  -- return the path components
+  link :: endpoint -> [String]
+
+instance HasLink api => HasLink (Static :> api) where
+  link (Static s :> api) = s : link api
+
+instance HasLink api => HasLink (Capture :> api) where
+  link (Capture :> api) = ??? : link api
+
+instance HasLink Verb where
+  link _ = []
+```
+
+We should be appending something in place of those `???` there.
+But since `Capture` represents variable path fragments (like `:userid` in
+`/user/:userid`, in many web frameworks), we do not want to pick a fixed string,
+we would like for this string to be supplied by the caller of `link`, as stated
+above. Let's introduce a slightly fancier `HasLink` class to make it seemingly
+"variadic".
+
+``` haskell
+class HasLink endpoint where
+  type LinkType endpoint :: *
+  link :: endpoint -> LinkType
+
+instance HasLink Verb where
+  type LinkType Verb = Link
+  link _ = []
+
+instance HasLink api => HasLink (Static :> api) where
+  type LinkType (Static :> api) = LinkType api
+  link (Static s :> api) = s : link api
```

Comment:
this doesn't work, does it? in real servant we have an accumulator of `Link` we fold the API over with.

Reply:
(here we already changed `link` type, it's not `Proxy api -> [String]`).

Reply:
Why wouldn't it? We're only describing a single endpoint at a time here, so we can just construct the path components as we go.

Reply:
Note that this is not the full servant approach here. It's not a `Proxy` that we take, the description is value+type level.

Reply:
e.g. `type API = Static :> Capture Int :> Verb`

```haskell
LinkType  (Static :> Capture  Int :> Verb) = String -> [String]

link :: (Static :> Capture  Int :> Verb) -> String -> [String]
link (Static s :> api) = s : link api -- non-sense!
```

Reply:
Well, you'd get an `Int` argument, but what's the problem with that?

Reply:
```haskell
{-# LANGUAGE TypeOperators, TypeFamilies #-}
{-# LANGUAGE FlexibleInstances #-}
-- static path fragments
data Static = Static String

-- variable path fragments ("captures")
data Capture = Capture

-- HTTP method
data Method = Get | Post
-- Leaf of a chain of :>'s, specifies the HTTP method
data Verb = Verb Method

-- chain a few "endpoint components" with this operator,
-- all chains must be terminated with a 'Verb' component.
infixr 5 :>
data a :> b = a :> b

-- a class to specify all the valid endpoint descriptions
class Endpoint a
instance Endpoint Verb
instance Endpoint rest => Endpoint (Static :> rest)
instance Endpoint rest => Endpoint (Capture :> rest)

-- GET /hello
endpoint1 = Static "hello" :> Verb Get


class HasLink endpoint where
  type LinkType endpoint :: *
--  link :: endpoint -> LinkType
  link :: endpoint -> LinkType endpoint

instance HasLink Verb where
--  type LinkType Verb = Link
  type LinkType Verb = [String]
  link _ = []

instance HasLink api => HasLink (Static :> api) where
  type LinkType (Static :> api) = LinkType api
  link (Static s :> api) = s : link api

{-
    • Couldn't match type ‘LinkType api’ with ‘[String]’
      Expected type: LinkType (Static :> api)
        Actual type: [String]
    • In the expression: s : link api

-}
```

Reply:
Oh dear, right. I'll fix this.

Reply:
I made the mistake a part of the post: https://github.com/haskell-servant/haskell-servant.github.io/pull/58/commits/a38ec1f2662987044415902bca703dffb9688839

Looks good?

Reply:
Good.